onnxruntime_extensions/tools/add_pre_post_processing_to_model.py

```diff
@@ -162,6 +162,134 @@ def superresolution(model_file: Path, output_file: Path, output_format: str, onn
+    You should assure this model has only one input, and the input shape is [1, 3, h, w].
+    The model has either one or more outputs. 
+        If the model has one output, the output shape is [1,num_boxes, coor+(obj)+cls] 
+            or [1, coor+(obj)+cls, num_boxes].
+        If the model has more than one outputs, you should assure the first output shape is 
+            [1, num_boxes, coor+(obj)+cls] or [1, coor+(obj)+cls, num_boxes].
+    Note: (obj) means it's optional.
+
+    :param model_file: The model file path.
+    :param output_file: The output file path.
+    """
+    model = onnx.load(str(model_file.resolve(strict=True)))
+    inputs = [create_named_value("image", onnx.TensorProto.UINT8, ["num_bytes"])]
+
+    model_input_shape = model.graph.input[0].type.tensor_type.shape
+    model_output_shape = model.graph.output[0].type.tensor_type.shape
+
+    # We will use the input_shape to create the model if provided by user.
+    if input_shape is not None:
+        assert len(input_shape) == 2, "The input_shape should be [h, w]."
+        w_in = input_shape[1]
+        h_in = input_shape[0]
+    else:
+        assert (model_input_shape.dim[-1].HasField("dim_value") and
+                model_input_shape.dim[-2].HasField("dim_value")), "please provide input_shape in the command args."
+
+        w_in = model_input_shape.dim[-1].dim_value
+        h_in = model_input_shape.dim[-2].dim_value
+
+    # Yolov5(v3,v7) has an output of shape (batchSize, 25200, 85) (Num classes + box[x,y,w,h] + confidence[c])
+    # Yolov8 has an output of shape (batchSize, 84,  8400) (Num classes + box[x,y,w,h])
+    # https://github.com/ultralytics/ultralytics/blob/e5cb35edfc3bbc9d7d7db8a6042778a751f0e39e/examples/YOLOv8-CPP-Inference/inference.cpp#L31-L33
+    # We always want the box info to be the last dim for each of iteration.
+    # For new variants like YoloV8, we need to add an transpose op to permute output back.
+    need_transpose = False
+
+    output_shape = [model_output_shape.dim[i].dim_value if model_output_shape.dim[i].HasField("dim_value") else -1
+                    for i in [-2, -1]]
+    if output_shape[0] != -1 and output_shape[1] != -1:
+        need_transpose = output_shape[0] < output_shape[1]
+    elif len(model.graph.input) > 1:
+        print(""" warning!!
+            The model has more than one inputs and the output shape is not fixed.
+            Output shape would be assumed to be [num_boxes, num_classes+4(reg)].
+            If this behavior is not expected, please run the onnx shape inference first to \
+            infer the output shape then run this script again.""")
+    else:
+        try:
+            import numpy as np
+            import onnxruntime
+        except ImportError:
+            raise ImportError(
+                "Please install onnxruntime and numpy to run this script. eg 'pip install onnxruntime numpy'")
```

Comment:
nit: tell the user why they need to do that - i.e. we need to execute the model to determine the output shape in order to add the correct post-processing.